experimental/dds/tree2/src/core/tree/anchorSet.ts

```diff
@@ -592,105 +591,239 @@ export class AnchorSet implements ISubscribable<AnchorSetRootEvents>, AnchorLoca
-				this.anchorSet.moveChildren(
-					sourcePath,
-					{ parent: this.parent, parentField: this.parentField, parentIndex: start },
+			beforeReplace(
+				newContent: FieldKey,
+				oldContent: Range,
+				destination: FieldKey,
+				kind: ReplaceKind,
+			): void {
+				assert(this.parentField !== undefined, "Must be in a field in order to replace");
+				const oldContentPath: RangeUpPath = {
+					parent: this.parent,
+					field: this.parentField,
+					...oldContent,
+				};
+				const newNodesSourcePath: DetachedRangeUpPath = brand({
+					field: newContent,
+					start: 0,
+					end: oldContent.end - oldContent.start,
+				});
+				const oldNodesDestinationPath: DetachedPlaceUpPath = brand({
+					field: destination,
+					index: 0,
+				});
+				for (const visitors of this.pathVisitors.values()) {
+					for (const pathVisitor of visitors) {
+						pathVisitor.beforeReplace(
+							newNodesSourcePath,
+							oldContentPath,
+							oldNodesDestinationPath,
+							kind,
+						);
+					}
+				}
+			},
+			afterReplace(
+				newContentSource: FieldKey,
+				newContent: Range,
+				oldContent: FieldKey,
+				kind: ReplaceKind,
+			): void {
+				assert(this.parentField !== undefined, "Must be in a field in order to replace");
+				const newContentPath: RangeUpPath = {
+					parent: this.parent,
+					field: this.parentField,
+					...newContent,
+				};
+				const newNodesSourcePath: DetachedPlaceUpPath = brand({
+					field: newContentSource,
+					index: 0,
+				});
+				const oldNodesDestinationPath: DetachedRangeUpPath = brand({
+					field: oldContent,
+					start: 0,
+					end: newContent.end - newContent.start,
+				});
+				for (const visitors of this.pathVisitors.values()) {
+					for (const pathVisitor of visitors) {
+						pathVisitor.afterReplace(
+							newNodesSourcePath,
+							newContentPath,
+							oldNodesDestinationPath,
+							kind,
+						);
+					}
+				}
+			},
+			replace(
+				newContentSource: FieldKey,
+				range: Range,
+				oldContentDestination: FieldKey,
+			): void {
+				this.notifyChildrenChanging();
+				this.detachEdit(range, oldContentDestination);
+				this.attachEdit(newContentSource, range.end - range.start, range.start);
+			},
+			destroy(detachedField: FieldKey, count: number): void {
+				this.anchorSet.removeChildren(
+					{
+						parent: undefined,
+						parentField: detachedField,
+						parentIndex: 0,
+					},
 					count,
 				);
 			},
+			beforeDestroy(detachedField: FieldKey, count: number): void {
+				const range: DetachedRangeUpPath = brand({
+					field: detachedField,
+					start: 0,
+					end: count,
+				});
+				for (const visitors of this.pathVisitors.values()) {
+					for (const pathVisitor of visitors) {
+						pathVisitor.beforeDestroy(range);
+					}
+				}
+			},
+			create(content: Delta.ProtoNodes, destination: FieldKey): void {
+				// Nothing to do since content can only be created in a new detached field,
+				// which cannot contain any anchors.
```

Comment:
Why doesn't this comment also apply to `afterCreate`?

Reply:
I think because `afterCreate` isn't creating any content, it's just notifing path visitors that creation is complete.